packages/replayio/src/utils/async/logProgress.ts

```diff
@@ -0,0 +1,78 @@
+import { dots } from "cli-spinners";
+import { disableAnimatedLog } from "../../config";
+import { logUpdate } from "../logUpdate";
+import { statusFailed, statusPending, statusSuccess } from "../theme";
+import { STATUS_PENDING, STATUS_REJECTED, STATUS_RESOLVED, Status } from "./createDeferred";
+
+export type LogProgressOptions = { delayBeforeLoggingMs?: number };
+
+export function logProgress(
+  initialMessage: string,
+  { delayBeforeLoggingMs = 0 }: LogProgressOptions = {}
+) {
+  let dotIndex = 0;
+  let logAfter = Date.now() + delayBeforeLoggingMs;
+  let status: Status = STATUS_PENDING;
+  let displayedMessage = initialMessage;
+
+  const print = () => {
+    let prefix: string;
+    switch (status) {
+      case STATUS_PENDING:
+        if (!disableAnimatedLog && delayBeforeLoggingMs > 0 && Date.now() < logAfter) {
+          return;
+        }
+
+        prefix = statusPending(dots.frames[++dotIndex % dots.frames.length]);
+        break;
+      case STATUS_REJECTED:
+        prefix = statusFailed("✘");
+        break;
+      case STATUS_RESOLVED:
+        prefix = statusSuccess("✔");
+        break;
+    }
+
+    if (displayedMessage) {
+      logUpdate(`${prefix} ${displayedMessage}`);
+    } else {
+      logUpdate.clear();
+    }
+  };
+
+  print();
+
+  const interval = disableAnimatedLog ? undefined : setInterval(print, dots.interval);
+
+  function assertPending() {
+    if (status !== STATUS_PENDING) {
+      throw Error(`logProgress is already in ${status} state`);
+    }
+  }
+
+  const finalize = () => {
+    clearInterval(interval);
+    print();
+    logUpdate.done();
+  };
+
+  return {
+    setFailed: (message: string) => {
+      assertPending();
+      status = STATUS_REJECTED;
+      displayedMessage = message;
+      finalize();
+    },
+    setProgress: (message: string) => {
```

Comment:
I think this should be "pending" to stay inline with other deferred statuses.
```suggestion
    setPending: (message: string) => {
```

IMO "progress" is a broader term that could also encompass "done" states